lib/hsa/mcwamp_hsa.cpp

```diff
@@ -3395,23 +3380,67 @@ class HSAContext final : public KalmarContext
         STATUS_CHECK(status, __LINE__);
 
-        // Iterate over agents to find out the first cpu device as host
-        status = hsa_iterate_agents(&HSAContext::find_host, &host);
-        STATUS_CHECK(status, __LINE__);
+        // Get a global memory pool from each CPU agent
+        // for the purpose of evaluating the NUMA distance
+        // to a GPU agent
+        std::vector<hsa_amd_memory_pool_t> cpu_mem_pools(cpu_agents.size());
+        auto find_global_mem_pool = [](hsa_amd_memory_pool_t pool, void* data) {
+            hsa_status_t status;
+
+            hsa_amd_memory_pool_global_flag_t flags;
+            status = hsa_amd_memory_pool_get_info(pool, 
+                                                  HSA_AMD_MEMORY_POOL_INFO_GLOBAL_FLAGS, 
+                                                  &flags);
+            STATUS_CHECK(status, __LINE__);
+
+            if ((flags & HSA_AMD_MEMORY_POOL_GLOBAL_FLAG_FINE_GRAINED)
+                || (flags & HSA_AMD_MEMORY_POOL_GLOBAL_FLAG_COARSE_GRAINED)) {
+                hsa_amd_memory_pool_t* global_mem_pool = static_cast< hsa_amd_memory_pool_t*>(data);
+                *global_mem_pool = pool;
+                return HSA_STATUS_INFO_BREAK;
+            }
+
+            return HSA_STATUS_SUCCESS;
+        };
+        for (int i = 0; i < cpu_agents.size(); ++i) {
+            hsa_amd_memory_pool_t global_mem_pool = {0};
+            status = hsa_amd_agent_iterate_memory_pools(cpu_agents[i],
+                                                        find_global_mem_pool,
+                                                        &global_mem_pool);
+            STATUS_CHECK(status, __LINE__);
+            cpu_mem_pools[i] = global_mem_pool;
+        }
 
         // The Devices vector is not empty here since CPU devices have
         // been added to this vector already.  This provides the index
         // to first GPU device that will be added to Devices vector
         int first_gpu_index = Devices.size();
-
-        Devices.resize(Devices.size() + agents.size());
-        for (int i = 0; i < agents.size(); ++i) {
-            hsa_agent_t agent = agents[i];
-            Devices[first_gpu_index + i] = new HSADevice(agent, host, i);
+        Devices.resize(Devices.size() + gpu_agents.size());
+        for (int i = 0; i < gpu_agents.size(); ++i) {
+            hsa_agent_t gpu_agent = gpu_agents[i];
+
+            // pick a CPU agent with the smallest NUMA distance
+            hsa_agent_t cpu_agent = {0};
+            uint32_t cpu_numa_distance = UINT32_MAX;
+            for (int j = 0; j < cpu_agents.size(); ++j) {
+
+              hsa_amd_memory_pool_link_info_t link_info = {0};
+              status = hsa_amd_agent_memory_pool_get_info(gpu_agent, 
+                                                          cpu_mem_pools[j],
+                                                          HSA_AMD_AGENT_MEMORY_POOL_INFO_LINK_INFO,
+                                                          &link_info);
```

Comment:
We need to figure out how many hops before querying for the link info.  This code here makes an assumption that there's only one hop